### ISSUE TITLE
varLib: Error Merging OpenType layout tables

### ISSUE THREAD

**Issue by eliheuer** [fonttools/fonttools#1353]
I'm trying to build a variable font from the [Cairo](https://github.com/Gue3bara/Cairo) sources with `fontmake  1.7.4`. Exporting from Glyphs works, but when trying to build with `fontmake -g sources/Cairo.glyphs -o variable`, I get the following error:

```
INFO:fontTools.varLib:Merging OpenType Layout tables
Traceback (most recent call last):
  File "/Users/Air/Py/Venvs/type/bin/fontmake", line 11, in <module>
    load_entry_point('fontmake', 'console_scripts', 'fontmake')()
  File "/Users/Air/Py/Source/fontmake/Lib/fontmake/__main__.py", line 258, in main
    project.run_from_glyphs(glyphs_path, **args)
  File "/Users/Air/Py/Source/fontmake/Lib/fontmake/font_project.py", line 561, in run_from_glyphs
    self.run_from_designspace(designspace_path, **kwargs)
  File "/Users/Air/Py/Source/fontmake/Lib/fontmake/font_project.py", line 646, in run_from_designspace
    **kwargs)
  File "/Users/Air/Py/Source/fontmake/Lib/fontmake/font_project.py", line 715, in run_from_ufos
    master_bin_dir=master_bin_dir)
  File "/Users/Air/Py/Source/fontmake/Lib/fontmake/font_project.py", line 268, in build_variable_font
    font, _, _ = varLib.build(designspace_path, finder)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 744, in build
    _merge_OTL(vf, model, master_fonts, axisTags)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 548, in _merge_OTL
    merger.mergeTables(font, master_fonts, ['GPOS'])
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 108, in mergeTables
    self.mergeThings(font[tag], [m[tag] for m in master_ttfs])
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 95, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 72, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 95, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 72, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 95, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 72, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/Air/Py/Venvs/type/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 99, in mergeThings
    assert all(out == v for v in lst), (out, lst)
AssertionError: ((3, [3, 5, 5]), 'int', '.FeatureCount', 'FeatureList', '.FeatureList', 'GPOS', '.table', 'table_G_P_O_S_')
```
I see there is a closed issue for a similar problem: https://github.com/fonttools/fonttools/issues/1125

**Comment by justvanrossum** [comment]
It means the number of GPOS features is not consistent across masters.

**Comment by khaledhosny** [comment]
This often means you have some glyphs with different anchors across masters.

**Comment by eliheuer** [comment]
Got it, thanks. I'll close this.

**Comment by eliheuer** [comment]
The problem here was beyond silly, one instance had a weight class that I didn't notice. Got it building now.

**Comment by m4rc1e** [comment]
@eliheuer could you specify exactly what was wrong?

Such info could make more meaningful error messages

**Comment by eliheuer** [comment]
One of the instances had a weight class custom parameter, I just didn't notice it. Hence this was a careless mistake on my part.

**Comment by davelab6** [comment]
Still the error should be better

**Comment by anthrotype** [comment]
the assertion error was originally reported was referring to the FeatureCount in the GPOS table, so it's unlikely that the issue (at least the one triggering that) was caused by differing weight class custom parameter (which if anything would affect the OS/2 table, not the OTL tables). The explanations from @justvanrossum and @khaledhosny are more likely to be correct.

**Comment by mjlagattuta** [comment]
I was having a similar issue, but it was resolved after fixing a poor anchor oversight of mine. This was the error I was getting:
```
AssertionError: ((4, [4, 3]), 'list', '.LigatureAttach', 'LigatureArray', '.LigatureArray', 'MarkLigPos', '[0]', 'Lookup', '[11]', 'list', '.Lookup', 'LookupList', '.LookupList', 'GPOS', '.table', 'table_G_P_O_S_')
```
I agree with @davelab6 that the error message could be better in pointing you to the glyphs and anchors that are problematic.

<br>

Another thing relevant to merging tables that I have come across when building [Inknut Antiqua](https://github.com/mjlagattuta/Inknut-Antiqua/blob/master/sources/InknutAntiqua.glyphs) is the following info message from fontmake:
```
INFO:fontTools.varLib:Merging OpenType Layout tables
INFO:fontTools.varLib:Generating gvar
INFO:fontTools.varLib:Merging TT hinting
INFO:fontTools.ttLib.tables.otBase:Attempting to fix OTLOffsetOverflowError ('GPOS', 'LookupIndex:', 8, 'SubTableIndex:', None, 'ItemName:', None, 'ItemIndex:', None)
INFO:fontTools.ttLib.tables.otBase:Attempting to fix OTLOffsetOverflowError ('GPOS', 'LookupIndex:', 11, 'SubTableIndex:', None, 'ItemName:', None, 'ItemIndex:', None)
```
I am curious if anyone knows what fontTools is "Attempting to fix" here or why there is an `OTLOffsetOverflowError` to begin with?

**Comment by davelab6** [comment]
I suggest making a calendar item to chat with Cosimo about this in December, he is away rest of the month

**Comment by thundernixon** [comment]
If anyone else comes across this, I was getting a similar error and solved it by making sure a bunch of things were similar between source fonts. In the end, the final thing I solved before it worked was setting the same glyph order for both UFOs. But before that, I also took a few other steps to check that there was a lot of similarity between anchors and kerning, so these steps may have also had an effect.

My main error message was this: 

```
AssertionError: ((2, [3, 2]), 'int', '.FeatureCount', 'FeatureList', '.FeatureList', 'GPOS', '.table', 'table_G_P_O_S_')
```

I wrote a bit more here: thundernixon/recursive#102
